src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionKey.cs

```diff
@@ -6,35 +6,41 @@ namespace System.Net.Http
 {
     internal readonly struct HttpConnectionKey : IEquatable<HttpConnectionKey>
     {
-        public readonly bool UsingSSL;
         public readonly string Host;
         public readonly int Port;
+        public readonly string SslHostName;     // null if not SSL
 
-        public HttpConnectionKey(Uri uri)
+        public HttpConnectionKey(string host, int port, string sslHostName)
         {
-            UsingSSL =
-                HttpUtilities.IsSupportedNonSecureScheme(uri.Scheme) ? false :
-                HttpUtilities.IsSupportedSecureScheme(uri.Scheme) ? true :
-                throw new ArgumentException(SR.net_http_client_http_baseaddress_required, nameof(uri));
-
-            Host = uri.Host;
-            Port = uri.Port;
+            Host = host;
+            Port = port;
+            SslHostName = sslHostName;
         }
 
+        // The common cases here are SslHostName == null or SslHostName == Host
         public override int GetHashCode() =>
```

Comment:
Q.: it is a trade-off and requires measurement for any informed decision, but, I expect the hash code to be used all the time so have you considered caching it and setting its value during construction? At 2 strings and 1 int, it seems that another int is still acceptable...

Reply:
Yeah, I thought about this a bit.  It's worth looking into more.